### PR TITLE
Add Unicode support and template expressions for strings 

### LINKED ISSUE #90: Lexical improvements
- unicode support for characters and strings
  - the typical approach for this is to include all unicode characters except the ones which need replacing
- unicode support for identifiers
  - use ID_START and ID_CONTINUE
- template strings
  - use the `$` syntax known from js an many other languages
  - in regular strings
  - make `$` escapable

### PR DESCRIPTION
**Pull request by nk-coding**
## technical changes
- update dependencies
- syntax highlighting now uses a type 2 grammar

## new features
- template expressions in strings
- unicode support for string literals and identifiers
- add comment support to textual editor (e.g. Ctrl + # for line comment)

## bug fixes
- lexer errors were not shown

## breaking changes
- `${` now is treated as a template in strings

## issues
- closes #90

### CODE REVIEW

File: packages/language-server/src/completion/executableCompletionExpression.ts [hunk 1]
```diff
@@ -242,23 +243,24 @@ export class ExecutableCompletionExpression extends ExecutableExpression<Express
...
+            } else if (match === "\t") {
+                return "\\t";
+            } else if (match === '"') {
+                return '\\"';
+            } else if (match === "\\") {
+                return "\\\\";
+            } else if (match === "${") {
+                return "\\${";
+            } else {
+                const code = match.charCodeAt(0);
```
[the remark this answers is not in the document]
Comment (nk-coding) [reply packages/language-server/src/completion/executableCompletionExpression.ts:260]: the characters we replace (based on the regex above) are all within this range

File: packages/language-server/src/completion/executableCompletionExpression.ts [hunk 2]
```diff
@@ -242,23 +243,24 @@ export class ExecutableCompletionExpression extends ExecutableExpression<Express
...
+            if (match === "\n") {
+                return "\\n";
+            } else if (match === "\t") {
+                return "\\t";
+            } else if (match === '"') {
+                return '\\"';
+            } else if (match === "\\") {
+                return "\\\\";
+            } else if (match === "${") {
+                return "\\${";
```
[the remark this answers is not in the document]
Comment (nk-coding) [reply packages/language-server/src/completion/executableCompletionExpression.ts:258]: you mean sth like this? `"Test\\\${"`
yes

File: packages/language-server/src/completion/executableCompletionExpression.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (nk-coding) [reply packages/language-server/src/completion/executableCompletionExpression.ts:260]: but maybe we want to add support for `\u{2-6 hex chars}` too?

File: packages/core/src/parser/lexer.ts [hunk 3]
```diff
@@ -11,7 +11,11 @@ export enum Modes {
     /**
      * Skip newlines in cst
      */
-    IGNORE_NEW_LINE = "IGNORE_NEW_LINES"
+    IGNORE_NEW_LINE = "IGNORE_NEW_LINES",
+    /**
+     * Mode inside
```
[the remark this answers is not in the document]
Comment (nk-coding) [reply packages/core/src/parser/lexer.ts]: added with 8511d8e1be8800f8e9ccb114dfe7d7a900144475

File: packages/language-server/src/completion/executableCompletionExpression.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (delvh) [reply packages/language-server/src/completion/executableCompletionExpression.ts:258]: Ah sorry, I meant accidental double escaping from the program.

File: packages/language-server/src/completion/executableCompletionExpression.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (delvh) [reply packages/language-server/src/completion/executableCompletionExpression.ts:260]: Ah, I misread the code.
We are inside the `replaceAll`.
It looked like we were outside of it.

[the remark this answers is not in the document]
Comment (delvh) [reply packages/language-server/src/completion/executableCompletionExpression.ts:260]: > but maybe we want to add support for \u{2-6 hex chars} too?

What do you mean?

File: packages/core/src/parser/lexer.ts [hunk 4]
```diff
@@ -13,7 +13,9 @@ export enum Modes {
      */
     IGNORE_NEW_LINE = "IGNORE_NEW_LINES",
     /**
-     * Mode inside
+     * Mode inside string literals
+     * Behaves significantly differnt from the other modes,
```
Comment (delvh) [inline packages/core/src/parser/lexer.ts:17]:
```suggestion
     * Behaves significantly differently from the other modes,
```

**Review by delvh — APPROVED** [review]

File: packages/language-server/src/completion/executableCompletionExpression.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (nk-coding) [reply packages/language-server/src/completion/executableCompletionExpression.ts:260]: currently, unicode escapes only work for characters in the BMP
if outside, you need to use a surrogate pair
js has the syntax `\u{xxxxxx}` so you can have such characters directly
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String#utf-16_characters_unicode_code_points_and_grapheme_clusters

**Comment by delvh** [comment]
Here are my thoughts on the implementation after testing:
It doesn't handle escapes well.

Input:
```scala
    a你们好 = "你们"
    if(a你们好 != null) {error("${a你们好}")}
```

Testing the different positions for `\` yields the following:
- `\${}`:
```scala
    a你们好 = "你们"
    if(a你们好 != null) {error("\${a你们好}")}
```
works as expected :heavy_check_mark: 
- `$\{}` doesn't work :x:, `$\\{}` however does :heavy_check_mark: 
- `${\}` doesn't work **:x:**, the `\` is recognized as escape symbol for the non-existing `\}` code
-  `${\\}` work at all **:x:`** . Even worse, its handling inside strings is even weirder:
```scala
    a你们好 = "你们"
    \\ = {args[0] + args[1]}
    if(a你们好 != null) {error("${a你们好 \\ "a"}}")}
```
works :heavy_check_mark: 
```scala
    a你们好 = "你们"
    \ = {args[0] + args[1]}
    if(a你们好 != null) {error("${a你们好 \\ "a"}}")}
```
which is the expected form however doesn't :x:

So, how do we want to handle all these edge cases?
I'd propose the following behavior: Inside

**Comment by delvh** [comment]
`println(${a你们好})` is understandably not allowed.
However, `println("${a你们好}")` is.
I can foresee a small amount of confusion for users who haven't worked with string interpolation so far.

File: packages/core/src/runtime/ast/executableStringLiteralExpression.ts [hunk 5]
```diff
@@ -32,7 +33,10 @@ export class ExecutableStringLiteralExpression extends ExecutableExpression<Stri
             } else {
                 const evaluationResult = part.expression.evaluate(context);
                 let evaluationResultValue = evaluationResult.value;
-                const toString = evaluationResultValue.getFieldValue("toString", context);
+                let toString: BaseObject | undefined;
+                if (!evaluationResultValue.isNull) {
+                    evaluationResultValue.getFieldValue("toString", context);
```
Comment (delvh) [inline packages/core/src/runtime/ast/executableStringLiteralExpression.ts:38]:
```suggestion
                    toString = evaluationResultValue.getFieldValue("toString", context);
```

**Pull request merged**